test/NuGet.Core.FuncTests/Dotnet.Integration.Test/ProjectFileBuilder.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            BaseDir = path;
+
+            fixture.CreateDotnetNewProject(path, ProjectName, " classlib");
+
+            ModifyProjectFile();
+        }
+
+        /// <summary>
+        /// Adds and intem inside a &lt;ItemGroup/&gt; node in the following form:
+        /// <c>&lt;{itemType} Include="{itemPath}" [PackagePath="{packagePath}"] /&gt;</c>
+        /// </summary>
+        public ProjectFileBuilder WithItem(string itemType, string itemPath, string packagePath)
+        {
+            ItemGroupEntries.Add(new ItemEntry 
+            {
+                ItemType = itemType, 
+                ItemPath = itemPath, 
+                PackagePath = packagePath,
+            });
+
+            return this;
+        }
+
+        public ProjectFileBuilder WithPackageIcon(string packageIcon)
+        {
+            PackageIcon = packageIcon;
+
+            return this;
+        }
+
+        public ProjectFileBuilder WithPackageIconUrl(string packageIconUrl)
+        {
+            PackageIconUrl = packageIconUrl;
+
+            return this;
+        }
+
+        public ProjectFileBuilder WithProjectName(string projectName)
+        {
+            ProjectName = projectName;
+
+            return this;
+        }
+
+        public ProjectFileBuilder WithProperty(string propertyName, string value)
+        {
+            Properties[propertyName] = value;
+
+            return this;
+        }
+
+        private void ModifyProjectFile()
+        {
+            using (FileStream stream = new FileStream(ProjectFilePath, FileMode.Open, FileAccess.ReadWrite))
+            {
+                var xml = XDocument.Load(stream);
+
+                if (PackageIconUrl != null)
+                {
+                    ProjectFileUtils.AddProperty(xml, "PackageIconUrl", PackageIconUrl);
+                }
+
+                if (PackageIcon != null)
+                {
+                    ProjectFileUtils.AddProperty(xml, "PackageIcon", PackageIcon);
+                }
+
+                ProjectFileUtils.AddProperties(xml, Properties);
+
+                var attributes = new Dictionary<string, string>();
+                var properties = new Dictionary<string, string>();
+                attributes["Pack"] = "true";
+                foreach (var tup in ItemGroupEntries)
+                {
+                    attributes.Remove("PackagePath");
+
+                    if (tup.PackagePath != null)
+                    {
+                        attributes["PackagePath"] = tup.PackagePath;
+                    }
+
+                    ProjectFileUtils.AddItem(xml, tup.ItemType, tup.ItemPath, string.Empty, properties, attributes);
+                }
+
+                ProjectFileUtils.WriteXmlToFile(xml, stream);
+            }
+        }
+    }
+
+    /// Represents an MsBuild project file item entry,
+    /// For testing purposes
+    public class ItemEntry
+    {
+        public string ItemType { get; set; }
```

Comment:
style: We should prefer immutable models wherever possible. 
They're considerably less error prone (even if this is only an inner class for a test).

Reply:
Ok. Will changet to read-only properties and a proper ctor.